pkg/agent/controller/traceflow/traceflow_controller.go

```diff
@@ -0,0 +1,391 @@
+	kubeClient            clientset.Interface
+	traceflowClient       clientsetversioned.Interface
+	traceflowInformer     opsinformers.TraceflowInformer
+	traceflowLister       opslisters.TraceflowLister
+	traceflowListerSynced cache.InformerSynced
+	ovsBridgeClient       ovsconfig.OVSBridgeClient
+	ofClient              openflow.Client
+	interfaceStore        interfacestore.InterfaceStore
+	networkConfig         *config.NetworkConfig
+	nodeConfig            *config.NodeConfig
+	queue                 workqueue.RateLimitingInterface
+	running               map[uint8]opsv1alpha1.Traceflow // tag->tf if tf.Status.Phase is Running.
+	senders               map[uint8]opsv1alpha1.Traceflow // tag->tf if this Node is sender.
+}
+
+func (c *Controller) TraceflowClient() clientsetversioned.Interface {
+	return c.traceflowClient
+}
+
+func (c *Controller) TraceflowInformer() opsinformers.TraceflowInformer {
+	return c.traceflowInformer
+}
+
+// NewTraceflowController instantiates a new Controller object which will process Traceflow
+// events.
+func NewTraceflowController(
+	kubeClient clientset.Interface,
+	traceflowClient clientsetversioned.Interface,
+	traceflowInformer opsinformers.TraceflowInformer,
+	client openflow.Client,
+	ovsBridgeClient ovsconfig.OVSBridgeClient,
+	interfaceStore interfacestore.InterfaceStore,
+	networkConfig *config.NetworkConfig,
+	nodeConfig *config.NodeConfig) *Controller {
+
+	running := make(map[uint8]opsv1alpha1.Traceflow)
+	senders := make(map[uint8]opsv1alpha1.Traceflow)
+
+	c := &Controller{
+		kubeClient:            kubeClient,
+		traceflowClient:       traceflowClient,
+		traceflowInformer:     traceflowInformer,
+		traceflowLister:       traceflowInformer.Lister(),
+		traceflowListerSynced: traceflowInformer.Informer().HasSynced,
+		ovsBridgeClient:       ovsBridgeClient,
+		ofClient:              client,
+		interfaceStore:        interfaceStore,
+		networkConfig:         networkConfig,
+		nodeConfig:            nodeConfig,
+		queue:                 workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "traceflow"),
+		running:               running,
+		senders:               senders}
+
+	// Add handlers for ClusterNetworkPolicy events.
+	traceflowInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addTraceflow,
+			UpdateFunc: c.updateTraceflow,
+			DeleteFunc: c.deleteTraceflow,
+		},
+		resyncPeriod,
+	)
+	return c
+}
+
+// enqueueTraceflow adds an object to the controller work queue.
+func (c *Controller) enqueueTraceflow(tf *opsv1alpha1.Traceflow) {
+	c.queue.Add(tf.Name)
+}
+
+// Run will create defaultWorkers workers (go routines) which will process the Traceflow events from the
+// workqueue.
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	// Traceflow is only available in Geneve tunnel.
+	if c.networkConfig.TunnelType != ovsconfig.GeneveTunnel {
```

Comment:
+ 1 to validate the config at agent startup.
@tnqn : a question - I thought feature gate is only for enabling/disabling a feature by user, but do you mean it can be used to control feature behavior due to other state/config options?

Reply:
It's indeed only for enabling/disabling a feature, but in our case this feature is only available when the config is specific value so I suggested to just disable the feature with a warning even if user enables it explicitly but doesn't configure the correct tunnel type.
I thought it again, maybe we shouldn't mix them as the feature will graduate from feature gate finally so it will need to check tunnel type directly anyway. Then to better interact with user, perhaps we should mark the traceflow as failed with a message like "Traceflow is only supported for tunnelType Geneve"

Reply:
Used with feature gate, and intra-node traceflow is supported when tunnel is not geneve.
Changed the code logic. If inter-node traceflow detected, Antrea Agent will validate tunnel mode and mark the traceflow as failed with a reason like "Error on node xxx" when tunnel is not geneve.